### DISCUSSION TITLE
Why I can't fetch the email/route.js ?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#67867] [Help] [answered]
### Summary

Error:
![image](https://github.com/user-attachments/assets/c428dd22-40d0-4404-a7a3-e822849d1618)

From: 
app/file/page.js

```
const response = await fetch("../my/email/route.js", {
        method: "POST",
        headers: {
          Accept: "application/json, text/plain, */*",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });

```
to:
app/my.email/route.js

**Comment by icyJoseph** [comment]
You need to use a relative path, to the API endpoint. The route.js is just a convention.

```js
fetch("/my/email")
```

Also, assuming that the route.js file implements a POST handler

**Comment by Elindo586** [comment]
New error is:
        
       `POST http://localhost:3000/my/email 405 (Method Not Allowed)`

What do you mean by assuming the route.js file implements a POST handler?

The POST is coming from  app/file/pages.js

Trying to communicate with:

app/my/email/route.js

Maybe I should upload it to Git, but this project is just to send emails from my computer, so it isn't meant to be a website to be hosted.

**Reply by Elindo586** [reply]
Oh this is weird...

This is the git for this project https://github.com/Elindo586/estuff/tree/main/app

It didn't seem to upload the "file" and "my" files..

![image](https://github.com/user-attachments/assets/be9f61f0-b98e-4e2f-8fda-60440fb6c27a)

**Reply by icyJoseph** [reply]
You need to `git add` the contents. Make sure the folders aren't empty.

What I meant with a POST handler is that you, literally need to have a function named POST in your route.js file, for example:


```js
export async function POST(request) {
  const res = await request.json()
  return Response.json({ res })
}
```

**Reply by Elindo586** [reply]
I did another `git add .` and now I got the files uploaded. 

If I change the name on top of the component to do an 
`export async function POST() {`

I get a new error:

`Error: The default export is not a React Component in page: "/file"`

It also won't allow me to wrap the fetch function as an export...  

I was able to upload the file here:
https://github.com/Elindo586/estuff/blob/main/app/file/page.js

**Reply by icyJoseph** [reply]
Ok, I think I am not being clear enough.

1. The POST handler needs to live in `my/email/route.js`
2. In app/file/page.js change `export async function POST()` to `export default function Page()`  
3. Back to `my/emai/route.js`:

```js
// my/email/route.js
export async function POST(req) {
```

And:

```js
  //  return res.status(200).json({ message: "Email sent!" }); remove this and add:
  return Response.json({ message: "Email sent!" });
```

To consume the body of the request:

```js
  const body =  await request.json()
  const nameFull = body.contact;
  const nameFirst = body.upper;
  const email = body.email;
  const title = body.id;
  const quote = body.quote;
```

Also, change this :` let nodemailer = require("nodemailer");` to


```js
import nodemailer from "nodemailer";

export function POST(req) {
```

**Reply by Elindo586** [reply]
It's sending an email, but not communicating to the front end properly.

The documentation shows an example of a fetch function inside a route.js file, but I would normally need the fetch in the front end to get data and send it out...  not sure what is typical to do ...

Now I am getting new errors.


From the Terminal:
```
email sent
 ⨯ Error: No response is returned from route handler 'C:\Users\Edgar\Desktop\Stuff\Acode\emails\app\my\email\route.js'. Ensure you return a `Response` or a `NextResponse` in all branches of your handler.
```

From the console in the browser:
![image](https://github.com/user-attachments/assets/bc77bc0e-97f6-4d5e-8bb7-28884778c1f2)


The files are re-uploaded.
https://github.com/Elindo586/estuff/tree/main/app

**Reply by icyJoseph** [reply, marked as answer]
Ok, again, confusion. 

Node.js also supports fetch, but in the server you can fetch from another services. Your frontend can fetch also other services, including your own backend. But this is for another day... I'd recommend taking your time with this new stack.

I see that you almost had it... the end of your `my/email/route.js` function is wrong:

```js
  await new Promise((resolve, reject) => {
    transporter.sendMail(mailData, function (err, info) {
      if (err) {
        console.log(err);
        reject(err);
      } else {
        console.log("email sent");
        resolve(info);
      }
    });
    return Response.json({ message: "Email sent!" }); // This is at the wrong place ❌
  });
}
```

Here, your project, constructor, returns the Response, but that's in the wrong place. Change it to:

```js
  await new Promise((resolve, reject) => {
    transporter.sendMail(mailData, function (err, info) {
      if (err) {
        console.log(err);
        reject(err);
      } else {
        console.log("email sent");
        resolve(info);
      }
    });
  });

 return Response.json({ message: "Email sent!" }); // After awaiting for the transporter return a response to the client
}
```

Now it is the POST function that returns a Response.

Oddly enough, this reveals a bug in your original function implementation (when it was Pages Router API handler). I don't want to confuse you, but I feel it is necessary to it point out. You were answering back `Email sent!`, always, and immediately, regardless of whether or not `transporter` has actually sent the email, without really waiting for it to complete.

That's because the Promise constructor, `new Promise((resolve, reject) => {/* this part */})`, runs synchronously, and you had your response sending bit, `Email sent!`, inside there! However, in Pages Router, calling `res.json` ends the request and sends the response, so you never really perceived a major issue. 

In App Router, the POST/GET/PUT, etc, handlers, have to return the response to the framework, and not doing so, stalls the request, and causes the issues you saw. Interesting.

**Reply by Elindo586** [reply]
OK I see it... 
`return Response.json({ message: "Email sent!" });`    has to be at the very end.
 
The back end must have` export async function POST(req)`  to handle the back end properly.

The front end seems almost with no changes other than  `"use client";`

There is one more error on the browser console, but this seems to be un-related to the emails.... I will open another post to check this one out.
